tensorboard/data/server/server.rs

```diff
@@ -820,6 +922,107 @@ mod tests {
         assert_eq!(xent_data.value, Vec::<f32>::new());
     }
 
+    #[tokio::test]
+    async fn test_list_tensors() {
+        let commit = CommitBuilder::new()
+            .run("run_with_no_data", None)
+            .scalars("train", "accuracy", |b| b.build())
+            .tensors("train", "weights", |b| b.build())
+            .tensors("test", "weights", |mut b| {
+                b.wall_time_start(1235.0).step_start(0).len(3).build()
+            })
+            .build();
+        let handler = sample_handler(commit);
+        let req = Request::new(data::ListTensorsRequest {
+            experiment_id: "123".to_string(),
+            plugin_filter: Some(data::PluginFilter {
+                plugin_name: "tensors".to_string(),
+            }),
+            ..Default::default()
+        });
+        let res = handler.list_tensors(req).await.unwrap().into_inner();
+
+        assert_eq!(res.runs.len(), 2);
+        let map = run_tag_map!(res.runs);
+
+        let test_run = &map[&Run("test".to_string())];
+        assert_eq!(test_run.len(), 1);
+        let weights_metadata = &test_run[&Tag("weights".to_string())]
+            .metadata
+            .as_ref()
+            .unwrap();
+        assert_eq!(weights_metadata.max_step, 2);
+        assert_eq!(weights_metadata.max_wall_time, 1237.0);
+        assert_eq!(
+            weights_metadata
+                .summary_metadata
+                .as_ref()
+                .unwrap()
+                .plugin_data
+                .as_ref()
+                .unwrap()
+                .plugin_name,
+            "tensors".to_string()
+        );
+    }
+
+    #[tokio::test]
+    async fn test_read_tensors() {
+        fn make_string_tensor_proto<T: Into<Bytes>>(value: T) -> pb::TensorProto {
```

Comment:
This is perfectly fine; it’s also idiomatic to write

```rust
fn make_string_tensor_proto(value: impl Into<Bytes>) -> ...
```

using an existential instead of a type parameter. These are mostly
equivalent; I don’t have a strong preference here.

Reply:
Done; thanks for the tip.